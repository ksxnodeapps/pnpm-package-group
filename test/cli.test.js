'use strict'
const {resolve, join} = require('path')
const {readFileSync, readdirSync, statSync} = require('fs')
const {spawnSync} = require('child_process')
const fsForce = require('fs-force')

const setOfLines = string => new Set(
  String(string)
    .split(/\n|\r/)
    .filter(Boolean)
    .map(x => x.trim())
)

describe('The command when everything aside itself works', () => {
  const outputDir = join(__dirname, 'cli-output.tmp')

  const expected = setOfLines(
    readFileSync(resolve(__dirname, 'data/cli-expect.txt'), 'utf8')
  )

  const testSuccess = description => {
    const received = setOfLines(
      spawnSync('node', [
        resolve(__dirname, '../index.js'),
        '--pnpm=echo',
        `--local=${outputDir}`,
        resolve(__dirname, 'data/cli-input.yaml')
      ]).stdout
    )

    describe(description, () => {
      it('should print correct messages', () => {
        expect(expected).toEqual(received)
      })

      describe('should create a folder tree', () => {
        describe('whose sub-items', () => {
          const names = readdirSync(outputDir)
          const paths = names.map(x => resolve(outputDir, x))

          it('have correct names', () => {
            expect(new Set(names)).toEqual(new Set(['local1', 'local2']))
          })

          it('are also folders', () => {
            paths.forEach(x =>
              expect(statSync(x).isDirectory()).toBe(true)
            )
          })
        })
      })
    })
  }

  fsForce.deleteSync(outputDir)
  testSuccess('when there was no output dir in the first place')
  fsForce.writeFileSync(outputDir, 'it is a file')
  testSuccess('when there output path was a file instead of a directory')
  testSuccess('when the directory is at right place')
})
