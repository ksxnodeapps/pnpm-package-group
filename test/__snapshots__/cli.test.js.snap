// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`program --help 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "$ index.js [options] [arguments]

Options:
  --version     Show version number                                    [boolean]
  --local, -l   Directory contains local packages         [string] [default: \\"\\"]
  --pnpm        Program to call                       [string] [default: \\"pnpm\\"]
  --shell       Whether shell is used                  [boolean] [default: true]
  --quiet-pnpm  Don't print pnpm output               [boolean] [default: false]
  --help        Show help                                              [boolean]

Examples:
  index.js packages.yaml                    Basic use
  index.js --local=local-packages           Specify target directory
  packages.yaml
  index.js packages.yaml/foo/bar            Install only group foo.bar in
                                            packages.yaml
  index.js /abc/def/packages.yaml/foo/bar   Combine filesystem path and json
                                            path
  index.js file1.yaml/foo/bar1              Filter files and groups
  file1.yaml/foo/bar2 file2.yaml/foo/bar1
  file2.yaml/foo/bar2
  index.js *.yaml/foo/{bar1,bar2}           Using blob pattern (UNIX Shell only)
",
}
`;

exports[`program being invoked without arguments 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "{ argv: 
   { _: [],
     version: false,
     shell: true,
     'quiet-pnpm': false,
     quietPnpm: false,
     help: false,
     local: '',
     l: '',
     pnpm: 'pnpm',
     '$0': 'index.js' } }",
}
`;
